src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+    {
+        cnsStr  = op1Str;
+        spanObj = op2;
+    }
+    else
+    {
+        cnsStr  = op2Str;
+        spanObj = op1;
+    }
+
+    int             cnsLength = -1;
+    const char16_t* str       = nullptr;
+    if (cnsStr->IsStringEmptyField())
+    {
+        // check for fake "" first
+        cnsLength = 0;
+        JITDUMP("Trying to unroll MemoryExtensions.Equals|SequenceEqual|StartsWith(op1, \"\")...\n", str)
+    }
+    else
+    {
+        str = info.compCompHnd->getStringLiteral(cnsStr->gtScpHnd, cnsStr->gtSconCPX, &cnsLength);
+        if (cnsLength < 0 || str == nullptr)
+        {
+            // We were unable to get the literal (e.g. dynamic context)
+            return nullptr;
+        }
+        JITDUMP("Trying to unroll MemoryExtensions.Equals|SequenceEqual|StartsWith(op1, \"%ws\")...\n", str)
+    }
+
+    CORINFO_CLASS_HANDLE spanCls      = gtGetStructHandle(spanObj);
+    CORINFO_FIELD_HANDLE pointerHnd   = info.compCompHnd->getFieldInClass(spanCls, 0);
+    CORINFO_FIELD_HANDLE lengthHnd    = info.compCompHnd->getFieldInClass(spanCls, 1);
+    const unsigned       lengthOffset = info.compCompHnd->getFieldOffset(lengthHnd);
+
+    // Create a placeholder for Span object - we're not going to Append it to statements
+    // in advance to avoid redundant spills in case if we fail to vectorize
+    unsigned spanObjRef         = lvaGrabTemp(true DEBUGARG("spanObjRef"));
+    lvaTable[spanObjRef].lvType = TYP_BYREF;
+
+    GenTree* spanObjRefLcl = gtNewLclvNode(spanObjRef, TYP_BYREF);
+    GenTree* spanLength    = gtNewFieldRef(TYP_INT, lengthHnd, gtClone(spanObjRefLcl), lengthOffset);
+    GenTree* spanData      = gtNewFieldRef(TYP_BYREF, pointerHnd, spanObjRefLcl);
+
+    // Additionally spill spanData to a local which we might need to clone then:
+    unsigned spanDataTmp         = lvaGrabTemp(true DEBUGARG("spanData tmp"));
+    lvaTable[spanDataTmp].lvType = TYP_BYREF;
+    GenTree* spanDataTmpLcl      = gtNewLclvNode(spanDataTmp, spanData->TypeGet());
+
+    GenTree* unrolled =
+        impExpandHalfConstEquals(spanDataTmpLcl, spanLength, false, startsWith, (WCHAR*)str, cnsLength, 0);
+    if (unrolled != nullptr)
+    {
+        // We succeeded, fill the placeholders:
+        impAssignTempGen(spanObjRef, impGetStructAddr(spanObj, spanCls, (unsigned)CHECK_SPILL_NONE, true));
+        impAssignTempGen(spanDataTmp, spanData);
+        if (unrolled->OperIs(GT_QMARK))
+        {
+            // QMARK can't be a root node, spill it to a temp
+            unsigned rootTmp = lvaGrabTemp(true DEBUGARG("spilling unroll qmark"));
+            impAssignTempGen(rootTmp, unrolled);
+            unrolled = gtNewLclvNode(rootTmp, TYP_INT);
+        }
+
+        JITDUMP("... Successfully unrolled to:\n")
+        DISPTREE(unrolled)
+
+        for (int i = 0; i < argsCount; i++)
+        {
+            impPopStack();
+        }
+
+        // We have to clean up GT_RET_EXPR for String.op_Implicit or MemoryExtensions.AsSpans
```

Comment:
This might conflict with my suggestion about using `gtRetExprVal` above -- if you do that you may not have an easy time finding the call.

Reply:
Yes, I added more asserts and comment to only handle single-level GT_RET_EXPR for now